File: R_sources/house_prices/attributes_selection.R

within(list(), 
{
    useless_numeric_vars <- c(   
        'BsmtFinSF2', 'BsmtHalfBath', 'BsmtUnfSF',
        'EnclosedPorch', 'KitchenAbvGr', 'LowQualFinSF',
        'MiscVal', 'PoolArea', 'ScreenPorch', 
        'WoodDeckSF', 'X3SsnPorch'
    )
    
    
    entropy <- within(list(), {
        
        entropy <- function(col) {
            counts = table(col, useNA="ifany")
            freqs <- counts / length(col)
            -sum(freqs * log2(freqs))
        }
        
        arrange_vars <- function(df) {
            df %>%
            purrr::map(entropy) %>% 
            as_data_frame %>% 
            gather(var, entropy) %>%
            arrange(entropy)
        }
    })
    
    
    groups_separation <- within(list(), {
        
        arrange_vars <- function (df, target_var) {
            
            target_var <- enquo(target_var)
            
            precalculated(df, !!target_var) %>%
            mutate(
                additive = (lead_mean - mean)^2 / (std^2 / freq + lead_std^2 / lead_freq)
            ) %>%
            summarise(
                score = sum(additive, na.rm = TRUE)
            ) %>%
            arrange(score)
        }
        
        precalculated <- function (df, target_var) {
            
            target_var <- enquo(target_var)
            target_var_char <- as.character(target_var)[2]

            global_std <- sd(df[,target_var_char][[1]], na.rm=TRUE)
            
            df %>%
            filter(!is.na(!!target_var)) %>%
            gather(var, value, -!!target_var) %>%
            group_by(var, value) %>%
            summarise(
                n = n(),
                mean = mean(!!target_var),
                std = ifelse(n > 1, sd(!!target_var), global_std)
            ) %>%
            mutate(
                freq = n / sum(n)
            ) %>% 
            arrange(var, mean) %>%
            mutate(
                lead_mean = lead(mean, 1),
                lead_freq = lead(freq, 1),
                lead_std  = lead(std, 1)
            )
        }
    })
    
    Rsquared <- within(list(), {
        
        arrange_vars <- function(df, target_var) {
            
            target_var <- enquo(target_var)
            target_var_char <- as.character(target_var)[2]
            
            formula <- reformulate(termlabels = c('value'), response = target_var_char)
            
            df %>%
            filter(!is.na(!!target_var)) %>%
            gather(var, value, -!!target_var) %>%
            group_by(var) %>%
            nest %>%
            mutate(
                mod = map(data, ~lm(formula, .)),
                glance = map(mod, broom::glance),
                r2 = map_dbl(glance, 'r.squared')
            ) %>%
            select(var, r2) %>%
            arrange(r2)
        }
    })
    
    
    anova <- within(list(), {
        
        Q_table <- function(df, target_var) {
            target_var <- enquo(target_var)
            target_var_char <- as.character(target_var)[2]
            
            df <- df %>% filter(!is.na(!!target_var))
            
            global_mean <- mean(df[,target_var_char][[1]])
            
            df %>% 
            gather(var, value, -!!target_var) %>%
            group_by(var, value) %>%
            summarise(
                n = n(),
                mean = mean(!!target_var),
                Q_within_group = sum( (`-`(!!target_var, mean))**2 ),
                Q_of_group = n * (mean - global_mean)**2,
                Q_total = sum( (`-`(!!target_var, global_mean))**2 )
            ) %>%
            group_by(var) %>%
            summarise(
                num_levels = n(),
                num_observ = sum(n),
                Q_within_groups = sum(Q_within_group),
                Q_of_groups = sum(Q_of_group),
                Q_total = sum(Q_total)
            )
        }
        
        arrange_vars <- function(df, target_var) {
            target_var <- enquo(target_var)
            
            Q_table(df, !!target_var) %>%
            mutate(
                F = (Q_of_groups * (num_observ - num_levels)) / (Q_within_groups * (num_levels - 1))
            ) %>%
            select(var, F) %>%
            arrange(F)
        }
    })
    
    
    order_factor_by_target <- function(df, factor_var, target_var, fn = mean) {
        factor_var <- enquo(factor_var)
        factor_var_char <- as.character(factor_var)[2]
        target_var <- enquo(target_var)

        factor_ordering <- 
            df %>%
            group_by(!!factor_var) %>%
            summarise(
                avg = fn(!!target_var)
            ) %>%
            arrange(avg) %>%
            select(!!factor_var) %>%
            `[[`(1)
        
        df %>%
        mutate(
            !!factor_var_char := factor(
                !!factor_var,
                levels=factor_ordering
            )
        )
    }
})

File: python_sources/kaggle/house_prices/attributes_selection.py

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf


def entropy(col):
    freqs = col.value_counts(dropna=False) / len(col)
    return -np.sum(freqs * np.log2(freqs))


def arrange_vars_by_entropy(df):
    return (
        df
        .apply(entropy)
        .sort_values()
        .to_frame('entropy')
        .reset_index()
        .rename(columns={"index": "var"})
    )


def arrange_vars_by_r2(df, target_var):
    return (
        pd.melt(
            frame=df.dropna(subset=[target_var]), 
            id_vars=[target_var], 
            var_name='var', 
            value_name='value'
        )
        .groupby('var')
        .apply(
            lambda df: smf.ols(
                formula='%s ~ value' % target_var, data=df
            ).fit().rsquared
        )
        .sort_values()
        .to_frame('r2')
        .reset_index()
    )


class GroupsSeparation:
    """ XXX """
    def __init__(self, df, target_var):
        self.df = df
        self.target_var = target_var
        self.df_long = pd.melt(
            frame=df.dropna(subset=[target_var]),
            id_vars=[target_var],
            var_name='var', 
            value_name='value'
        )

    def arrange_vars(self):
        precalculated = self.get_precalculated()
        return (
            precalculated
            .assign(additive=lambda df: self.calc_additive(df))
            .groupby(['var'])
            ['additive']
            .sum()
            .sort_values()
            .to_frame('score')
            .reset_index()
        )
   
    def get_precalculated(self):
        global_std = self.df[self.target_var].std()

        summ = (
            self.df_long.groupby(['var', 'value'], as_index=False)
                [self.target_var]
                .agg({
                    'n': np.count_nonzero,
                    'mean': np.mean,
                    'std': lambda vec: self.calc_std(vec, global_std)
                })
                .assign(
                    freq=lambda this: (this.groupby('var')
                                           ['n']
                                           .apply(lambda n: n / np.sum(n)))
                )
                .sort_values(['var', 'mean'])
        )

        grp = summ.groupby(['var'], group_keys=False)
        summ['lead_mean'] = grp.apply(lambda this: this['mean'].shift(-1))
        summ['lead_freq'] = grp.apply(lambda this: this['freq'].shift(-1))
        summ['lead_std'] = grp.apply(lambda this: this['std'].shift(-1))

        return summ

    @staticmethod
    def calc_std(vec, unit_default=0):
        if len(vec) == 0:
            return 0
        if len(vec) == 1:
            return unit_default
        if len(vec) > 1:
            return np.std(vec, ddof=1)
    
    @staticmethod    
    def calc_additive(df):
        up = (df['lead_mean'] - df['mean'])**2
        down = df['std']**2 / df['freq'] + df['lead_std']**2 / df['lead_freq']
        return up / down


class Anova:
    """ XXX """
    def __init__(self, df, target_var):
        self.df = df.dropna(subset=[target_var])
        self.target_var = target_var

    def Q_table(self):
        global_mean = np.mean(self.df[self.target_var])
        return (
            pd.melt(
                frame=self.df,
                id_vars=[self.target_var],
                var_name='var', 
                value_name='value'
            )
            .groupby(['var', 'value'], as_index=False)
            [self.target_var]
            .agg({
                'n': np.count_nonzero,
                'mean': np.mean,
                'Q_within_group': lambda vec: np.sum((vec - np.mean(vec))**2),
                'Q_total': lambda vec: np.sum((vec - global_mean)**2)
            })
            .assign(
                Q_of_group=lambda df: df['n'] * (df['mean'] - global_mean)**2
            )
            .groupby(['var'])
            .apply(lambda grp: pd.Series({
                'num_levels': grp.shape[0],
                'num_observ': np.sum(grp['n']),
                'Q_within_groups': np.sum(grp['Q_within_group']),
                'Q_of_groups': np.sum(grp['Q_of_group']),
                'Q_total': np.sum(grp['Q_total']),
            }))
            .reset_index()
        )

    def arrange_vars(self):
        q_table = self.Q_table()

        def calc_F(df):
            up = (df['Q_of_groups'] * (df['num_observ'] - df['num_levels']))
            down = (df['Q_within_groups'] * (df['num_levels'] - 1))
            return up / down 
        
        return (
            q_table
            .assign(
                F=calc_F
            )
            [['var', 'F']]
            .sort_values('F')
        )

    


def order_factor_by_target(df, factor_var, target_var, fn=np.mean):
    """ XXX """
    factor_ordering = (
        df
        .groupby(factor_var, as_index=False)
        .agg(fn)
        .sort_values(target_var)
        [factor_var]
    )

    return (
        df
        .assign(
            **{factor_var: (
                lambda df: df[factor_var]
                       .astype('category')
                       .cat.reorder_categories(factor_ordering, ordered=True))
        })
    )    
